### ISSUE TITLE
Include all pages in template context, not just current page.

### ISSUE THREAD

**Issue by engfer** [mkdocs/mkdocs#1119]
The templating system and HTML generation method assume that the page is the always the scope of generation. This prevents single-page HTML themes, which requires contents from **all** pages, from being generated without some magics (workarounds) to get around the generation system. 

I know that [mkdocs-pandoc](https://github.com/jgrassler/mkdocs-pandoc) essentially does this by walking the TOC/pages, but as a template designer, it's quite limiting.

Please let me know if I am missing something.

**Comment by engfer** [comment]
This feature could also be the path to alleviate everyone's needs for printing or writing the entire docs to PDF

**Comment by waylan** [comment]
@engfer you spell out what you don't like well enough. But how would you like it to be instead. Sorry, the way it works now seems natural to me. Perhaps I have never needed your use case, so its not obvious to me.

**Comment by d0ugal** [comment]
I think @engfer wants the ability to do something like this...

```
{% for page in pages %}
    {{ page.title }}
    {{ page.content }}
{% endfor %}
```

However, at the moment, we never pass context with all pages to the template. We only ever pass the current page. This is something I attempted to tackle in the past, with the in-memory markdown processing that loaded all the pages rather than one at a time but I abandoned the work due to issues with it and lack of time. The branch may still be floating around somewhere...

**Comment by d0ugal** [comment]
I think this was it: https://github.com/d0ugal/mkdocs/tree/refactor-markdown-loading

**Comment by waylan** [comment]
> I think this was it: https://github.com/d0ugal/mkdocs/tree/refactor-markdown-loading

Ahh, yes I remember that. In fact, with my recent work on the Plugin API I was wishing for that (having meta available before converting Markdown to HTML would make for a better API). Although, by storing the Markdown content for every page in memory before rendering any one, we should have some sort of file-based cache for the page instances (IIRC, that is exactly what Pelican (a static blogging engine) does). Otherwise those really large projects (thousands of pages) will run into problems (we already get reports from those people for the large search index so I know they exist). 

So, yes, I want to add this. But it is harder than it may seem on the surface. Not sure if this should be a 1.0 feature or post-1.0.

**Comment by waylan** [comment]
This has been addressed in #1180.
